### PR TITLE
WebGPURenderer: support clipping

### PR DESCRIPTION
**Pull request by aardgoose**
Supports global and material bases clipping.

Notes:

This can be extended to support  a ClippingGroup() as mentioned at:

https://github.com/mrdoob/three.js/blob/a2bf250301187940fd97f60d5c020691ad3b8c11/src/renderers/WebGLRenderer.js#L1783-L1786

Fixes shadow handling of shadow side.

Added support for uniform arrays for vec4 only. Is this method ok or is a full set of <type>ArrayUniform and <type>ArrayNodeUniform classes prefered?

### CODE REVIEW

File: examples/jsm/nodes/accessors/ClippingNode.js [hunk 1]
```diff
@@ -0,0 +1,210 @@
...
+		return this.clippingNode;
+
+	}
+
+	setupAlphaToCoverage() {
+
+		const numClippingPlanes = this._numClippingPlanes;
+		const numUnionClippingPlanes = this._numUnionClippingPlanes;
+
+		this.clippingNode = tslFn( () => {
```
Comment (LeviPesin) [inline examples/jsm/nodes/accessors/ClippingNode.js:108]: Is `clippingNode` needed as a property, couldn't it just be returned from `setup`?

File: examples/jsm/renderers/webgpu/WebGPURenderer.js [hunk 2]
```diff
@@ -23,7 +23,7 @@ class WebGPURenderer extends Renderer {
 
 		let BackendClass;
 
-		if ( parameters.forceWebGL ) {
+		if ( ! parameters.forceWebGL ) {
```
Comment (LeviPesin) [inline examples/jsm/renderers/webgpu/WebGPURenderer.js:26]: Please revert this.

**Comment by LeviPesin** [comment]
> Added support for uniform arrays for vec4 only. Is this method ok or is a full set of ArrayUniform and ArrayNodeUniform classes prefered?

I guess we should support arbitrary types.

File: examples/jsm/nodes/accessors/ClippingNode.js (hunk 1 above)
Comment (aardgoose) [reply examples/jsm/nodes/accessors/ClippingNode.js:108]: Fixed now

Comment (LeviPesin) [reply examples/jsm/nodes/accessors/ClippingNode.js:108]: No, I've actualy meant cleaning this up in ClippingNode 😅 But the cleanup in NodeMaterial also seems nice to me.

File: examples/jsm/renderers/webgpu/WebGPURenderer.js (hunk 2 above)
Comment (aardgoose) [reply examples/jsm/renderers/webgpu/WebGPURenderer.js]: testing line got left in.

**Comment by aardgoose** [comment]
> > Added support for uniform arrays for vec4 only. Is this method ok or is a full set of ArrayUniform and ArrayNodeUniform classes prefered?
> 
> I guess we should support arbitrary types.

I was wondering what method to use as opposed to which types.  This approach just detects an array 'value' in the Uniform code and UniformGroups then handles arrays as part of the existing path.  Or should NodeBuilder allocate a 'Vec{n}ArrayNodeUniform' etc which requires more boilerplate classes. There is an ArrayUniformNode too, but that doesn't appear to be used anywhere yet, and this works with a standard uniform() node.

**Comment by sunag** [comment]
> There is an ArrayUniformNode too, but that doesn't appear to be used anywhere yet, and this works with a standard uniform() node.

`ArrayUniformNode` was initially implemented here ( https://github.com/mrdoob/three.js/pull/22497 ) but there is no related example anymore. I was also implementing array logic for floats, we will certainly help each other with that.

**Comment by sunag** [comment]
Maybe we need to add `webgpu_clipping` example in puppeteer too?

**Comment by aardgoose** [comment]
This now uses uniforms().

I'm not sure what is going on with the screen shot failure. The major area of difference is the bottom edge which isn't subject to clipping.

**Comment by aardgoose** [comment]
We commented at the same time.
I'll exclude it from the tests.

**Pull request merged**